src/bk-user/bkuser/plugins/ldap/client.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+"""
+TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-用户管理(Bk-User) available.
+Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
+Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
+You may obtain a copy of the License at http://opensource.org/licenses/MIT
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+"""
+
+from typing import Dict, List
+
+from ldap3 import ALL_ATTRIBUTES, DEREF_NEVER, SAFE_SYNC, Connection, Server
+from ldap3.extend.standard.PagedSearch import paged_search_accumulator
+
+from bkuser.plugins.ldap.constants import REQUIRED_OPERATIONAL_ATTRIBUTES
+from bkuser.plugins.ldap.exceptions import DataNotFoundError
+from bkuser.plugins.ldap.models import ServerConfig
+
+
+class LDAPClient:
+    """LDAP 客户端"""
+
+    def __init__(self, server_config: ServerConfig):
+        self.server_config = server_config
+        self._conn = self._gen_conn(server_config)
+        self._conn.bind()
+
+    def __del__(self):
```

Comment:
重写 `__del__` 是我们目前的插件规范的一部分吗？不是的话，不太建议去重写它，很容易出现一些和 GC 相关的奇奇怪怪的问题。以前使用 kubernetes SDK 等，遇到过不少。

FYI: [Cleanup ThreadPool with atexit rather than \_\_del\_\_ by fabianvf · Pull Request #1073 · kubernetes-client/python](https://github.com/kubernetes-client/python/pull/1073)

Reply:
有考虑过这种写法，但是这样的话，client 就难以初始化后，多处使用了，只能用的时候 with 一下

---
不过生成下 Client 的成本也不高，可以改

Reply:
我觉得没事，大部分 Python SDK 的 client 现在都是支持 with，以及要求用户总是手动 close 的。